Tools/AP_Periph/rangefinder.cpp

```diff
@@ -31,67 +34,92 @@ void AP_Periph_FW::can_rangefinder_update(void)
+        if (backend == nullptr) {
+            continue;
+        }
+
+        RangeFinder::Status status = backend->status();
+        if (status <= RangeFinder::Status::NoData) {
+            // don't send any data for this instance
+            continue;
+        }
 
-    canard_broadcast(UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_SIGNATURE,
-                    UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_ID,
-                    CANARD_TRANSFER_PRIORITY_LOW,
-                    &buffer[0],
-                    total_size);
+        const uint32_t sample_ms = backend->last_reading_ms();
+        if (last_rangefinder_sample_ms[instance] == sample_ms) {
+            // don't same the same reading again
+            continue;
+        }
+        last_rangefinder_sample_ms[instance] = sample_ms;
+
+        uavcan_equipment_range_sensor_Measurement pkt {};
+        pkt.sensor_id = rangefinder.get_address(instance);
+
+        switch (status) {
+        case RangeFinder::Status::OutOfRangeLow:
+            pkt.reading_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_READING_TYPE_TOO_CLOSE;
+            break;
+        case RangeFinder::Status::OutOfRangeHigh:
+            pkt.reading_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_READING_TYPE_TOO_FAR;
+            break;
+        case RangeFinder::Status::Good:
+            pkt.reading_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_READING_TYPE_VALID_RANGE;
+            break;
+        default:
+            pkt.reading_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_READING_TYPE_UNDEFINED;
+            break;
+        }
+
+        switch (backend->get_mav_distance_sensor_type()) {
+        case MAV_DISTANCE_SENSOR_LASER:
+            pkt.sensor_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_SENSOR_TYPE_LIDAR;
+            break;
+        case MAV_DISTANCE_SENSOR_ULTRASOUND:
+            pkt.sensor_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_SENSOR_TYPE_SONAR;
+            break;
+        case MAV_DISTANCE_SENSOR_RADAR:
+            pkt.sensor_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_SENSOR_TYPE_RADAR;
+            break;
+        default:
+            pkt.sensor_type = UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_SENSOR_TYPE_UNDEFINED;
+            break;
+        }
+
+        float dist_m = backend->distance();
+        pkt.range = dist_m;
+
+        uint8_t buffer[UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_MAX_SIZE] {};
+        uint16_t total_size = uavcan_equipment_range_sensor_Measurement_encode(&pkt, buffer, !periph.canfdout());
+
+        canard_broadcast(UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_SIGNATURE,
+                        UAVCAN_EQUIPMENT_RANGE_SENSOR_MEASUREMENT_ID,
+                        CANARD_TRANSFER_PRIORITY_LOW,
+                        &buffer[0],
+                        total_size);
+
+        rangefinder_last_sent_index = instance;
+        last_rangefinder_update_ms = now;
+        break;
```

Comment:
comment saying we only send one at a time

Reply:
better to just send all at the same time

Reply:
Done